### PR TITLE
Add signal resampling GPU kernel

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:

 **New feature**

## Description:

Adds a GPU signal resampling kernel, to be used in a standalone GPU resampling operator.

### Affected modules and functionalities:

Modified signal resampling a little bit to accommodate for the GPU version

### Key points relevant for the review:

Kernel efficiency, tests

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-1445

### CODE REVIEW

File: dali/kernels/signal/resampling.h [hunk 1]
```diff
@@ -54,19 +55,19 @@ inline float32x4_t vsetq_f32(float x0, float x1, float x2, float x3) {
 #endif
 
 struct ResamplingWindow {
-  inline std::pair<int, int> input_range(float x) const {
+  inline DALI_HOST_DEV ivec<2> input_range(float x) const {
```
Comment (mzient) [inline dali/kernels/signal/resampling.h:58]: I gather that's because `std::pair` doesn't work with device code? I can't say I'm very happy with (ab)using a vector as start/end.

File: dali/kernels/signal/resampling.h [hunk 2]
```diff
@@ -297,8 +308,16 @@ struct Resampler {
...
+        if (downmix) {
+          float out_val = 0;
+          for (int c = 0; c < num_channels; c++)
+            out_val += tmp[c];
+          out_val /= num_channels;
+          out[out_pos] = ConvertSatNorm<Out>(out_val);
+        } else {
+          for (int c = 0; c < num_channels; c++)
+            out[out_pos * num_channels + c] = ConvertSatNorm<Out>(tmp[c]);
+        }
```
Comment (mzient) [inline dali/kernels/signal/resampling.h:320]: This is _extremely_ wasteful. We should have some temporary buffer and downmix there, then resample, like we do in the decoder. In the long run that buffer can be smaller and we can downmix in chunks. I'd prefer not to have fused downmixing than have it done this way. Anyway, downmixing is cheap, we can keep it in the decoder.

File: dali/kernels/signal/resampling_gpu.cuh [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  int64_t out_block = static_cast<int64_t>(blockIdx.x) * blockDim.x;
+  int64_t start_out_pos = out_block + threadIdx.x;
+
+  double in_block_f = out_block * scale;
+  int64_t in_block_i = std::floor(in_block_f);
+  float in_pos_start = in_block_f - in_block_i;
+  const In* in_blk_ptr = in + in_block_i * nchannels;
+
+  for (int64_t out_pos = start_out_pos; out_pos < sample.out_len;
+       out_pos += grid_stride, in_pos_start += fscale * grid_stride) {
```
Comment (mzient) [inline dali/kernels/signal/resampling_gpu.cuh:63]: You're losing precision here. The CPU code advanced the position by one (scaled) sample. The position was calculated relative to the start of a block, thereby avoiding loss of precision. The number of times this position was advanced was limited to block_size and the maximum offset was block_size * scale. Here, you advance an unknown number of times and the range can span entire recording. This defeats the whole purpose of using double precision in the first place. The options are:
change the traversal order, so that a thread block processes adjacent blocks of data:
```
Block index for each output element - block size is 4, grid size is 3
[0 0 0 0 1 1 1 1 2 2 2 2 0 0 0 0 1 1 1 1 2 2 2 2 0 0 0 0]

Proposed order:
[ 0 0 0 0  0 0 0 0  1 1 1 1  1 1 1 1  1 1 1 1  2 2 2 2  2 2 2 2]
```
Also, we could try to measure the penalty of having the double precision calculations in each loop. Perhaps we can calculate the double precision numbers only in thread 0 of a warp and then use warp shuffling - this should help on machines with reduced fp64 performance (GTX/RTX/Titan).

Comment (mzient) [reply dali/kernels/signal/resampling_gpu.cuh:63]: This a serious problem, not a "nice to have" - this makes it impossible to address individual source samples in recordings longer than 9 minutes at 16 kHz. Between 5 and 9 minutes it would decay to nearest neighbor resampling (yikes!) and 9 to 17 minutes we'd only be able to address even samples, after 17 minutes every fourth and so on. You get the picture.

Comment (jantonguirao) [reply dali/kernels/signal/resampling_gpu.cuh]: The problem should be resolved with the latest version

File: dali/kernels/signal/resampling_gpu.cuh [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    return ConvertSatNorm<float>(in_val);  // just normalize
+  }
+}
+
+/**
+ * @brief Resamples 1D signal (single or multi-channel), optionally converting to a different data type.
+ *
+ * @param samples sample descriptors
+ */
+template <typename Out, typename In, bool SingleChannel = false>
```
Comment (awolant) [inline dali/kernels/signal/resampling_gpu.cuh:60]: I'm not asking to split it now, but for my taste the `SingleChannel` optimization could have been done in a separate PR.

File: dali/kernels/signal/resampling.h [hunk 5]
```diff
@@ -54,19 +59,23 @@ inline float32x4_t vsetq_f32(float x0, float x1, float x2, float x3) {
...
+
+  inline DALI_HOST_DEV InputRange input_range(float x) const {
     int xc = std::ceil(x);
     int i0 = xc - lobes;
     int i1 = xc + lobes;
     return {i0, i1};
   }
 
-  inline float operator()(float x) const {
+  inline DALI_HOST_DEV float operator()(float x) const {
```
Comment (awolant) [inline dali/kernels/signal/resampling.h:73]: Maybe this deserves a comment or a doc to explain why it is this way. For me it was not clear from the code, why it is done this way. I know you did not change how it works, but since you are touching it :)

File: dali/kernels/signal/resampling_gpu.cuh [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  int64_t grid_stride = static_cast<int64_t>(gridDim.x) * blockDim.x;
+  int64_t out_block_start = sample.out_begin + static_cast<int64_t>(blockIdx.x) * blockDim.x;
+
+  for (int64_t out_pos = out_block_start + threadIdx.x; out_pos < sample.out_end;
+       out_block_start += grid_stride, out_pos += grid_stride) {
+    double in_block_f = out_block_start * scale;
+    int64_t in_block_i = std::floor(in_block_f);
+    float in_pos_start = in_block_f - in_block_i;
+    const In* in_blk_ptr = in + in_block_i * nchannels;
+    float in_pos = in_pos_start + fscale * threadIdx.x;
```
Comment (awolant) [inline dali/kernels/signal/resampling_gpu.cuh:90]: I would be in favor of adding the explanation of the block/thread setup as a comment here, since Michał already have done a nice visualization of it in the review comment. Both how it is and why it is this way as the why fact is rather obscure in my opinion.

**Review by awolant — APPROVED** [review]

File: dali/kernels/signal/resampling_gpu.cuh (hunk 6 above)
Comment (jantonguirao) [reply dali/kernels/signal/resampling_gpu.cuh:90]: Actually we ended up with the original layout, but calculating `in_block_f` inside the loop, since the precision problem could be noticeable with either layout. I'll try to add an explanation on what's going on here in the comments

File: dali/kernels/signal/resampling_gpu.cuh [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  int64_t grid_stride = static_cast<int64_t>(gridDim.x) * blockDim.x;
+  int64_t out_block_start = sample.out_begin + static_cast<int64_t>(blockIdx.x) * blockDim.x;
+
+  for (int64_t out_pos = out_block_start + threadIdx.x; out_pos < sample.out_end;
+       out_block_start += grid_stride, out_pos += grid_stride) {
+    // A floating point distance `in_pos_start` is calculated from an arbitrary relative
+    // position, keeping the floats small in order to keep precision. `in_block_f`, used to
+    // calculate the reference for distance (in_block_i) needs to be calculated in double precision.
+    double in_block_f = out_block_start * scale;
+    int64_t in_block_i = floorf(in_block_f);
```
Comment (mzient) [inline dali/kernels/signal/resampling_gpu.cuh:90]:
```suggestion
    int64_t in_block_i = floor(in_block_f);
```

File: dali/kernels/signal/resampling_gpu.cuh [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_RESAMPLING_GPU_CUH_
+#define DALI_KERNELS_SIGNAL_RESAMPLING_GPU_CUH_
+
+#include <cuda_runtime.h>
+#include "dali/kernels/signal/resampling.h"
+#include "dali/core/util.h"
+
+#define SHM_NCHANNELS 16
```
Comment (mzient) [inline dali/kernels/signal/resampling_gpu.cuh:22]:
```suggestion
constexpr int SHM_NCHANNELS = 16;
```
(possibly with different capitalization)?

File: dali/kernels/signal/resampling_test.cc [hunk 9]
```diff
@@ -16,93 +16,149 @@
...
+      out_sh.tensor_shape_span(s)[1] = nchannels;
+    }
+  }
+  ttl_in_.reshape(in_sh);
+  ttl_out_.reshape(out_sh);
+  ttl_outref_.reshape(out_sh);
+  for (int s = 0; s < nsamples; s++) {
+    double in_rate = args[s].in_rate;
+    double out_rate = args[s].out_rate;
+    double scale = static_cast<double>(in_rate) / out_rate;
```
Comment (mzient) [inline dali/kernels/signal/resampling_test.cc:57]:
```suggestion
    double scale = in_rate / out_rate;
```
in_rate is already `double`.

File: dali/kernels/signal/resampling_gpu.cuh (hunk 8 above)
Comment (jantonguirao) [reply dali/kernels/signal/resampling_gpu.cuh:22]: this gives me this error:
```error: identifier "dali::kernels::signal::resampling::kShmChannels" is undefined in device code```

File: dali/kernels/signal/resampling_test.cc (hunk 9 above)
Comment (jantonguirao) [reply dali/kernels/signal/resampling_test.cc]: Fixed

**Comment by jantonguirao** [comment]
!build

File: dali/kernels/signal/resampling_gpu.cuh (hunk 8 above)
Comment (mzient) [reply dali/kernels/signal/resampling_gpu.cuh:22]: The problem lies with `cuda_min`, which takes a reference. A workaround would be:
`enum : int { kShmChannels = 16 };` - I'm not pushing it, just posting it here for completeness.

File: dali/kernels/signal/resampling_gpu.h [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+      out_sh[0] = out_end - out_begin;
+    }
+    req.output_shapes = {out_shape};
+    return req;
+  }
+
+  void Run(KernelContext &context, const OutListGPU<OutputType> &out,
+           const InListGPU<InputType> &in, span<const Args> args) {
+    if (window_gpu_storage_.empty())
+      Initialize();
```
Comment (mzient) [inline dali/kernels/signal/resampling_gpu.h:70]: There's a bug - the H2D copy in Initialize happens on stream 0 and may not complete before the processing starts here.

**Comment by jantonguirao** [comment]
!build

File: dali/kernels/signal/resampling_test.h [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+ protected:
+  int nsamples_ = 1;
+  int nchannels_ = 1;
+  double default_freq_in_ = 0.1;
+  int nsec_ = 1;
+  float eps_ = 2e-3;
+  float max_avg_err_ = 1e-3;
+  bool use_envelope_ = true;
+  int64_t roi_start_ = 0;
+  int64_t roi_end_ = -1;  // means end-of-signal
```
Comment (mzient) [inline dali/kernels/signal/resampling_test.h:70]: Nitpick: strictly speaking, this isn't your typical ROI, which is defined in terms of input.

Comment (mzient) [inline dali/kernels/signal/resampling_test.h:70]: Nitpick:
```suggestion
  int64_t out_roi_start_ = 0;
  int64_t out_roi_end_ = -1;  // means end-of-signal
```
Typically, a region of interest is defined in terms of the input.

File: dali/kernels/signal/resampling_test.cc [hunk 12]
```diff
@@ -16,93 +16,184 @@
...
+  this->roi_end_ = 8000;
+  this->nchannels_ = 8;
+  this->RunTest();
+}
+
+TEST_F(ResamplingCPUTest, SingleChannelNeedHighPrecision) {
+  this->default_freq_in_ = 0.49;
+  this->nsec_ = 400;
+  this->roi_start_ = 4000000;  // enough to look at the tail
+  this->roi_end_ = -1;
```
Comment (mzient) [inline dali/kernels/signal/resampling_test.cc:183]: use_envelope_ = false?

File: dali/kernels/signal/resampling_gpu_test.cu [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+TEST_F(ResamplingGPUTest, PerfTest) {
+  this->RunPerfTest(1000);
+}
+
+TEST_F(ResamplingGPUTest, SingleChannelNeedHighPrecision) {
+  this->default_freq_in_ = 0.49;
+  this->nsec_ = 400;
+  this->roi_start_ = 4000000;  // enough to look long into the signal
```
Comment (mzient) [inline dali/kernels/signal/resampling_gpu_test.cu:141]: Wasn't it meant to have `use_envelope_ = false`?

**Review by mzient — APPROVED** [review]
